feat!: migrate to Entity Framework Core 6.0

olavloite opened the pull request:
Migrates the provider to EF Core 6.0.

Updates #83

jskeet commented on Google.Cloud.EntityFrameworkCore.Spanner.IntegrationTests/QueryTests.cs:
```diff
@@ -555,7 +555,7 @@ public async Task CanUseStringConcat()
             var calculatedFullName = await db.Singers
                 .Where(s => new long[] { singerId }.Contains(s.SingerId))
-                .Select(s => s.FirstName + " " + s.LastName)
+                .Select(s => string.Concat(s.FirstName, " ", s.LastName))
```
Why do we need this change, given that we've got a test that shows we can use the + operator for strings?

olavloite replied:
Oops. Seems like a left-over from the initial development, when I didn't bother to create separate tests for plus and concat. I've added a separate test for the plus operator.

(The other tests use a mock server, these tests are actual integration tests)

jskeet commented on Google.Cloud.EntityFrameworkCore.Spanner.IntegrationTests/SpannerTestDatabase.cs:
```diff
@@ -118,8 +120,15 @@ private async Task MaybeCreateInstanceAsync()
+                EmulatorDetection = EmulatorDetection.EmulatorOrProduction,
+                GrpcAdapter = GrpcNetClientAdapter.Default,
             };
+            if (Environment.GetEnvironmentVariable("SPANNER_EMULATOR_HOST") != null)
```
Note that in the next major version of GAX and the Spanner libraries (coming soon), Grpc.Net.Client will be the default gRPC implementation - this code will need to change.

(I'm assuming the new Spanner version will come out before the EFCore 6 support... I really think it's worth waiting for that.)

olavloite replied:
What is 'soon'?. I would prefer that this PR does not drag on for too long, as the time needed to get deeply invested in this again in let's say a month is relatively high. Changing this (test) code a little later should be relatively easy.

Or will the new GAX and Spanner libraries require changes to the 'normal' usage of the libraries as well?

jskeet replied:
FYI: Google.Cloud.Spanner.Data (etc) are now being released as GA, version 4.0.0. (By the time you read this, the release process is likely to have completed.) So now would be a good time to update.

jskeet commented on Google.Cloud.EntityFrameworkCore.Spanner/Migrations/Internal/SpannerRelationalAnnotationProvider.cs:
```diff
@@ -0,0 +1,94 @@
+namespace Google.Cloud.EntityFrameworkCore.Spanner.Migrations.Internal
+{
+    /// <summary>
+    ///     This is internal functionality and not intended for public use.
```
Nit: odd leading whitespace here.

olavloite replied:
Removed

jskeet commented on Google.Cloud.EntityFrameworkCore.Spanner/Migrations/Internal/SpannerRelationalAnnotationProvider.cs:
```diff
@@ -0,0 +1,94 @@
+        {
+        }
+
+        public override IEnumerable<IAnnotation> For(IColumn column, bool designTime)
```
Even though this is all expected for internal use, I'd still expect XML comments for all publicly-accessible members.

jskeet reviewed (commented):
I'm not really qualified to review this, I'm afraid - I'll add Rishabh from the LEX team, who is likely to be doing a lot of the Spanner work for that team moving forward.

Rishabh-V commented:
Thanks Jon.

I will be happy to take a look (with my limited understanding of library I may add), but will be able to reach to it sometime in second half of next week. Thanks.

olavloite commented:
> Thanks Jon.
> 
> I will be happy to take a look (with my limited understanding of library I may add), but will be able to reach to it sometime in second half of next week. Thanks.

Hi @Rishabh-V Any update on this?

Rishabh-V commented:
> > Thanks Jon.
> > I will be happy to take a look (with my limited understanding of library I may add), but will be able to reach to it sometime in second half of next week. Thanks.
> 
> Hi @Rishabh-V Any update on this?

Hey Knut,

My apologies. I couldn't get to it yet. The earliest I can get to it now should be coming Thursday, if that is fine. Sorry for the delay.

Rishabh-V commented on Google.Cloud.EntityFrameworkCore.Spanner/Migrations/Internal/SpannerRelationalAnnotationProvider.cs:
```diff
@@ -0,0 +1,97 @@
+﻿// Copyright 2020, Google Inc. All rights reserved.
```
nit: 2022?

olavloite replied:
Done

Rishabh-V commented on Google.Cloud.EntityFrameworkCore.Spanner.Tests/Migrations/MigrationSqlGeneratorTestBase.cs:
```diff
@@ -758,13 +760,19 @@ protected virtual void Generate(params MigrationOperation[] operation)
+            modelBuilder.Entity<Albums>();
             modelBuilder.Model.RemoveAnnotation(CoreAnnotationNames.ProductVersion);
-
+            
```
nit: odd whitespaces?

olavloite replied:
Fixed

Rishabh-V reviewed (commented):
My apologies for getting to it this late. I had a look today and the code changes look ok to me. Most of the changes are in line with changes for EF Core upgrade and loose coupling. Unfortunately, with little exposure and understanding of the library, I cannot do contextual justice to the review with reference to the library.

amanda-tarafa commented on .github/workflows/build-and-test.yml:
```diff
@@ -14,7 +14,7 @@ jobs:
       uses: actions/setup-dotnet@v1
       with:
-        dotnet-version: 3.1.x
+        dotnet-version: 6.0.x
```
Will the netcoreapp3.1 have testing as well, both through actions and CI? It probably should if you'll continue to support it by updating dependencies and bug fixes.

olavloite replied:
We should set that up specifically for the v3.1 branch. The v6.0 Entity Framework provider will not have support for .NET 3.1.

amanda-tarafa reviewed (commented):
I've no knowledge at all of the specificities of EF Core 6, but this looks good overall.

nagytech commented:
Facing the 3.1 end of LTS issue myself I updated my version simply by bumping the version in each .csproj.  Everything works fine for me with minimal changes.  Granted there are additional changes to get the CI/CD system working, this should be at most a few lines of diff.

This PR on the other hand contains a wide array of trivial and unnecessary changes that could break a feature that someone had relied on previously.  Not to mention, reformatting and adding sugar in certain areas makes back-patching a bit hard.

I'm all for pushing this forward so we can all get a new package, but the changes here should be separated into two categories:

- Things that are necessary for upgrading to .NET; and 
- "Things I thought I would clean up while I was here."

olavloite commented:
@jskeet @amanda-tarafa @Rishabh-V I've updated everything to the latest Spanner client (and also the other dependencies) and addressed the other review comments. PTAL.

jskeet commented on Google.Cloud.EntityFrameworkCore.Spanner.IntegrationTests/Google.Cloud.EntityFrameworkCore.Spanner.IntegrationTests.csproj:
```diff
@@ -17,8 +19,9 @@
   <ItemGroup>
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Proxies" Version="3.1.25" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.25">
+    <PackageReference Include="Google.Api.Gax.Grpc.GrpcNetClient" Version="3.7.0" />
```
You shouldn't need this at all now (and it could cause issues).

olavloite replied:
Removed

jskeet reviewed (commented):
I don't have the resources to do a full review of this I'm afraid, but I've pointed out one minor issue.

amanda-tarafa reviewed (commented):
Except for Jon's comments, this looks good as far as sanity checking goes.

olavloite commented:
> In terms of upgrading to the next .NET version, there are a lot of unnecessary updates in this PR.
> 
> It would be a lot easier to review, and we might have a release by now if the auxiliary changes were pulled out.

This PR contains two (three) large updates:
1. It updates the .NET Entity Framework (Core) version from 3.1 to 6. This also means that the .NET version changes from 3.1 to 6.
2. It updates the Spanner data client from 3.x to 4.x.

The above is done to get a version that is linked to the most recent, LTS versions of the two main dependencies of the provider. That unfortunately also means that it is a large change.

> there are a lot of unnecessary updates in this PR.

I don't think there are. Aside from the occasional removal of some warnings, all changes in this PR are related to the two changes above. This also includes a lot of updates to the test cases, some which may seem unnecessary at first, but are down to for example small modifications in the format of the SQL that is generated by Entity Framework in 3.1 and 6.x.

AlexBagnolini commented:
@olavloite what is the next step in this?

Seems like Rishabh did review and approve the changes, but is still marked as "requested for review" for this to merge in.

Thanks!

olavloite commented:
> @olavloite what is the next step in this?
> 
> Seems like Rishabh did review and approve the changes, but is still marked as "requested for review" for this to merge in.
> 
> Thanks!

Hi @AlexBagnolini Sorry for the slow reply. I have a large amount of tasks on my plate at the moment, but I'll try to get to this somewhere during this week. As far as I'm concerned, the next steps will be:
1. Fix the merge conflicts and make sure everything is up to date and that all checks are green.
2. Have a quick sync with the reviewers to see if anyone has any final thoughts, and hopefully get a formal approval so we can merge it. (Merging is only possible with an actual review approval, that is currently missing)

ansh0l reviewed (approved):
Adding approval for signoff, since the PR is already reviewed with multiple other folks.
